File: implement-shell-tools/ls/myls.js
import { program } from "commander";
import { promises as fs } from "node:fs";
import process from "node:process";

program
  .name("myls")
  .description("list file(s) in the directory, like the ls command")
  .option("-1", "list one file per line")
  .option("-a", "include hidden files")
  .argument("[directory]", "Directory to list", ".");

program.parse();

const options = program.opts();
const directory = program.args[0] || ".";

try {
  let files = await fs.readdir(directory);

  // if "-a" is used, include hidden files; those that start with "." 
  if (options.a) {
    files = [".", "..", ...files];
  }

  for (const file of files) {
    // if "-a" is not used, skip hidden files; those that start with "."
    if (!options.a && file.startsWith(".")) {
      continue;
    }

    console.log(file); // print file name; one file per line
  }
} catch (error) {
  console.error(`Error reading directory ${directory}:`, error.message);
  process.exit(1);
}

File: implement-shell-tools/ls/myls.py
import argparse
import os

parser = argparse.ArgumentParser(
    prog="myls",
    description="List directory contents like the ls command"
)

parser.add_argument("-1", dest="one_per_line", action="store_true",
                    help="List one file per line")
parser.add_argument("-a", dest="all_files", action="store_true",
                    help="Include hidden files starting with '.'")
parser.add_argument("path", nargs="?", default=".", help="Directory to list (default: current directory)")

args = parser.parse_args()

try:
    entries = os.listdir(args.path)
except FileNotFoundError:
    print(f"my_ls: cannot access '{args.path}': No such file or directory")
    exit(1)

if not args.all_files:
    entries = [e for e in entries if not e.startswith('.')]

entries.sort()

if args.one_per_line:
    for entry in entries:
        print(entry)
else:
    print("  ".join(entries))
